### ISSUE TITLE
Unable to load modules on aarch64 linux machine via a rust based host app

### ISSUE THREAD

**Issue by rbackhouse** [bytecodealliance/wasmtime#4972]
Thanks for filing a bug report! Please fill out the TODOs below.

### Steps to Reproduce

* The app uses the following to load the module
```
let mut config = wasmtime::Config::new();
let engine = wasmtime::Engine::new(&config)?;
let module = wasmtime::Module::from_file(&engine, "./test.wasm")?;
```
### Expected Results

The module should load 

### Actual Results

Load fails with 

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 22, kind: InvalidInput, message: "Invalid argument" }', /home/rbackhouse/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmtime-jit-1.0.0/src/code_memory.rs:63:14

### Versions and Environment

Wasmtime version or commit: 1.0.0

Operating system: Ubuntu 18.04

Architecture: aarch64

### Extra Info

The CPU being used is a Quad-core ARM Cortex-A57 MPCore processor.

Looking at code_memory.rs line 60 it seems the call 
```
rustix::process::membarrier(
                rustix::process::MembarrierCommand::RegisterPrivateExpeditedSyncCore,
            )
            .unwrap();
```
is failing

If I comment out this call and the one at line 168 then everything works fine.

**Comment by alexcrichton** [comment]
cc @akirilov-arm

**Comment by bjorn3** [comment]
Which kernel version are you using? And which glibc version?

**Comment by rbackhouse** [comment]
Kernel version is 4.9.253-tegra
glibc version is (Ubuntu GLIBC 2.27-3ubuntu1.4) 2.27

**Comment by bjorn3** [comment]
MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED_SYNC_CORE has been introduced in linux 4.16. Maybe wasmtime will have to fall back to MEMBARRIER_CMD_SHARED when MEMBARRIER_CMD_PRIVATE_EXPEDITED_SYNC_CORE is not available?

**Comment by akirilov-arm** [comment]
I am not sure if `MEMBARRIER_CMD_SHARED` would provide the same guarantee as `MEMBARRIER_CMD_PRIVATE_EXPEDITED_SYNC_CORE`, while it definitely seems more heavyweight, since it affects all threads in the system (not just the ones in the current process). IMHO the first decision to make is how old of a kernel version should Wasmtime support.

**Comment by akirilov-arm** [comment]
OK, I got confirmation that `MEMBARRIER_CMD_SHARED` was an acceptable substitute, keeping in mind the following:
* It is a much more expensive operation because it is affected by all threads in all processes in the system, instead of only the threads in the current process that have registered explicitly for the notification
* It could still fail if the `nohz_full` parameter is passed to the kernel

**Comment by akirilov-arm** [comment]
`MEMBARRIER_CMD_SHARED` is [an alias](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/include/uapi/linux/membarrier.h?h=v5.19.12#n158) of `MEMBARRIER_CMD_GLOBAL` for the sake of backward compatibility.

**Comment by cfallin** [comment]
Ah! In that case, let's see if #4987 works (@rbackhouse would you be able to test this patch?).

**Comment by rbackhouse** [comment]
@cfallin yes I can try it

**Comment by rbackhouse** [comment]
@cfallin 
crates/jit/src/code_memory.rs line 168 also has a call to rustix::process::membarrier.

Should this have the same 

.or_else(|_| rustix::process::membarrier(rustix::process::MembarrierCommand::Global))

added too ?

**Comment by akirilov-arm** [comment]
@cfallin Note that the `cranelift-jit` crate has [similar code paths](https://github.com/bytecodealliance/wasmtime/blob/c1d6ca48a74ab6f356d08baa533b90b17ccb834a/cranelift/jit/src/backend.rs#L466), though currently they appear to fail silently.

**Comment by bxq2011hust** [comment]
I have met the same issue.

**Comment by afonso360** [comment]
I'm not sure that is the same issue, you have kernel 4.18 which should support this! Or fail with code `22` if it isn't implemented on AArch64.

However I did a quick google search and it might be that [the `membarrier` syscall isn't whitelisted on docker](https://stackoverflow.com/a/60495279/2065450), can you check if that's whats happening?

**Comment by bxq2011hust** [comment]
@afonso360  Yes, you are right, I found the docker engine version is 1.13. Upgrade docker to 20.10 and the test is fine. Thank you very much!

**Comment by alexcrichton** [comment]
I think this has since been resolved, so I'm going to close this.
